netty/src/main/java/io/grpc/netty/NettyAdaptiveCumulator.java

```diff
@@ -204,16 +204,9 @@ static void mergeWithCompositeTail(
       composite.addFlattenedComponents(true, newTail);
       // New tail's ownership transferred to the composite buf.
       newTail = null;
-      in.release();
-      in = null;
-      // Restore the reader. In case it fails we restore the reader after releasing/forgetting
-      // the input and the new tail so that finally block can handles them properly.
       composite.readerIndex(prevReader);
+      in.release(); // Successful so took over ownership of in
```

Comment:
```suggestion
      // Input buffer was successfully merged with the tail.
      // Must be the last line in the try because we release it only on success.
      in.release();
```